beacon_node/execution_layer/src/engine_api/http.rs

```diff
@@ -216,6 +224,115 @@ impl EngineApi for HttpJsonRpc {
     }
 }
 
+pub struct BuilderHttpJsonRpc(HttpJsonRpc);
+
+impl BuilderHttpJsonRpc {
+    pub fn new(url: SensitiveUrl) -> Result<Self, Error> {
+        HttpJsonRpc::new(url).map(Self)
+    }
+
+    pub async fn rpc_request<T: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: serde_json::Value,
+        timeout: Duration,
+    ) -> Result<T, Error> {
+        self.0.rpc_request(method, params, timeout).await
+    }
+}
+
+#[async_trait]
+impl EngineApi for BuilderHttpJsonRpc {
+    async fn upcheck(&self) -> Result<(), Error> {
+        self.0.upcheck().await
+    }
+
+    async fn get_block_by_number<'a>(
+        &self,
+        query: BlockByNumberQuery<'a>,
+    ) -> Result<Option<ExecutionBlock>, Error> {
+        self.0.get_block_by_number(query).await
+    }
+
+    async fn get_block_by_hash<'a>(
+        &self,
+        block_hash: ExecutionBlockHash,
+    ) -> Result<Option<ExecutionBlock>, Error> {
+        self.0.get_block_by_hash(block_hash).await
+    }
+
+    async fn new_payload_v1<T: EthSpec>(
+        &self,
+        execution_payload: ExecutionPayload<T>,
+    ) -> Result<PayloadStatusV1, Error> {
+        self.0.new_payload_v1(execution_payload).await
+    }
+
+    async fn get_payload_v1<T: EthSpec, Txns: Transactions<T>>(
+        &self,
+        payload_id: PayloadId,
+    ) -> Result<ExecutionPayload<T, Txns>, Error> {
+        self.0.get_payload_v1(payload_id).await
+    }
+
+    async fn forkchoice_updated_v1(
+        &self,
+        forkchoice_state: ForkChoiceState,
+        payload_attributes: Option<PayloadAttributes>,
+    ) -> Result<ForkchoiceUpdatedResponse, Error> {
+        self.0
+            .forkchoice_updated_v1(forkchoice_state, payload_attributes)
+            .await
+    }
+
+    async fn exchange_transition_configuration_v1(
+        &self,
+        transition_configuration: TransitionConfigurationV1,
+    ) -> Result<TransitionConfigurationV1, Error> {
+        self.0
+            .exchange_transition_configuration_v1(transition_configuration)
+            .await
+    }
+}
+
+#[async_trait]
+impl BuilderApi for BuilderHttpJsonRpc {
+    async fn get_payload_header_v1<T: EthSpec, Txns: Transactions<T>>(
+        &self,
+        payload_id: PayloadId,
+    ) -> Result<ExecutionPayload<T, Txns>, Error> {
+        let params = json!([JsonPayloadIdRequest::from(payload_id)]);
+
+        let response: JsonExecutionPayloadHeaderV1<T, Txns> = self
+            .rpc_request(
+                BUILDER_GET_PAYLOAD_HEADER_V1,
+                params,
+                BUILDER_GET_PAYLOAD_HEADER_TIMEOUT,
+            )
+            .await?;
+
+        Ok(response.into())
+    }
+
+    async fn propose_blinded_block_v1<T: EthSpec>(
+        &self,
+        block: SignedBeaconBlock<T, BlindedTransactions>,
+    ) -> Result<ExecutionPayload<T>, Error> {
+        let blinded_block = SignedBlindedBeaconBlock::from(block);
```

Comment:
The issue I ran into was the name of the field `transactions` (signed block's execution payload) vs `transactions_root` (signed blinded block's execution payload header). We could maybe figure out a way to do a conditional rename but I don't think there's a way to do it within `serde_derive`.  You can [alias](https://serde.rs/field-attrs.html#alias) on deserialization, but there's nothing like a `rename_if` for serialization.

This is actually making me realize we'd be slightly off-spec as-is because our BN<>VC communication would have the field name `transactions` if we don't do this same transformation in that API.

Reply:
I've opened a PR to this branch that uses `serde(flatten)` to unify the types, let me know if you think it's flexible enough: https://github.com/realbigsean/lighthouse/pull/20